Add AutoWeightsLoader utility for simplified weight loading

adityakamat24 opened the pull request:
Addresses #11864

<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

This PR adds `AutoWeightsLoader` utility class to simplify model weight loading code, addressing issue #11864.

  Currently, each model in sglang implements its own `load_weights()` method with repetitive logic for:
  - Iterating through weights
  - Skipping unused weights (rotary_emb.inv_freq, etc.)
  - Handling stacked parameters
  - Loading parameters with custom weight loaders

  This leads to code duplication across 100+ model files. The `AutoWeightsLoader` provides a reusable utility that
  abstracts this common logic, similar to vLLM's implementation.

## Modifications

**Added `AutoWeightsLoader` class to `python/sglang/srt/models/utils.py`:**

  - **Core functionality**: Automatically detects child modules and parameters while iterating weights only once
  - **Flexible filtering**:
    - `skip_prefixes`: Skip weights by prefix (e.g., "lm_head")
    - `skip_substrs`: Skip weights containing substrings (e.g., "rotary_emb.inv_freq")
    - `ignore_unexpected_prefixes/suffixes`: Ignore missing weights without errors
  - **Custom loading support**:
    - Respects `weight_loader` attributes on parameters
    - Calls custom `load_weights()` methods on submodules
  - **Auto-skip**: Automatically skips common unused weights (rotary_emb.inv_freq, cos_cached, sin_cached)
  - **Batch norm support**: Handles non-parameter tensors like batch normalization statistics

## Accuracy Tests

 This PR does not affect model outputs. It provides a utility class for weight loading that models can optionally adopt. No changes to existing models, kernels, or forward passes.

  The implementation maintains the same weight loading behavior as manual implementations:
  - Uses the same `default_weight_loader` function
  - Respects custom `weight_loader` attributes
  - Preserves parameter shapes and values

## Benchmarking and Profiling

This PR does not impact inference speed. It only affects the model initialization/loading phase, not the forward pass. Weight loading happens once at startup and is not performance-critical.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).


## Usage Example

  ```python
  from sglang.srt.models.utils import AutoWeightsLoader

  class MyModel(nn.Module):
      def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
          skip_prefixes = []
          if self.config.tie_word_embeddings:
              skip_prefixes.append("lm_head")

          loader = AutoWeightsLoader(
              self,
              skip_prefixes=skip_prefixes,
              skip_substrs=["projector"],  # Skip vision projectors
              ignore_unexpected_suffixes=[".attn.bias"],  # Ignore attention masks
          )
          return loader.load_weights(weights)

Commit:
Add AutoWeightsLoader utility for simplified weight loading

  - Add AutoWeightsLoader class to sglang.srt.models.utils
  - Implements automatic module/parameter detection during weight loading
  - Supports skip_prefixes, skip_substrs for filtering weights
  - Supports ignore_unexpected_prefixes/suffixes for optional weights
  - Auto-skips common unused weights (rotary_emb.inv_freq, etc.)
  - Respects custom weight_loader methods on parameters
  - Supports nested modules with custom load_weights methods
  - Reduces code duplication across model implementations

  Addresses #11864

Commit:
Fix AutoWeightsLoader code review issues

  - Create defensive copies of input lists to prevent mutation
  - Fix type hint: change nn.Parameter to torch.Tensor in _load_param
  - Remove unused Callable import

Commit:
Apply black-jupyter formatting

b8zhong commented:
Can you show the result of a test?

yhyang201 commented:
Great job! 

You can try using your `AutoWeightsLoader` with one of the models (I think `Qwen2` would be a good choice — see this test file for reference: [test_qwen_models.py](https://github.com/sgl-project/sglang/blob/main/test/srt/models/test_qwen_models.py).
That way, when CI runs, it will actually be testing your implementation.

You could also share your local test results from that file directly — that would be perfectly fine too.

Commit:
Test AutoWeightsLoader with Qwen2 model

  - Modified Qwen2ForCausalLM.load_weights() to use AutoWeightsLoader
  - Maintains backward compatibility by handling stacked params and PP manually
  - Delegates standard weight loading to AutoWeightsLoader for testing
  - Part of PR #12534 for AutoWeightsLoader implementation

adityakamat24 commented:
> Great job!
> 
> You can try using your `AutoWeightsLoader` with one of the models (I think `Qwen2` would be a good choice — see this test file for reference: [test_qwen_models.py](https://github.com/sgl-project/sglang/blob/main/test/srt/models/test_qwen_models.py). That way, when CI runs, it will actually be testing your implementation.
> 
> You could also share your local test results from that file directly — that would be perfectly fine too.

I've tested the `AutoWeightsLoader` implementation by integrating it with the Qwen2 model as suggested. Here are the results:

Results:
Accuracy: 0.820
Invalid: 0.000
Latency: 17.518 s
Output throughput: 1662.416 token/s
metrics={'accuracy': np.float64(0.82), 'invalid': np.float64(0.0), 'latency': 17.518477192148566, 'output_throughput': 1662.4161838137593}
PASSED

Changes Made:
Modified `python/sglang/srt/models/qwen2.py` to use `AutoWeightsLoader` for standard weight loading while maintaining
compatibility with complex cases (stacked parameters, pipeline parallelism). 

Would you like me to include this Qwen2 integration in the PR, or keep it as a separate follow-up?

Commit:
finalize and added AutoWeightsLoader unit tests

  - Reverted Qwen2 AutoWeightsLoader integration
  - Keep feature separate from testing
  - Unit test provides CI coverage (per-commit-1-gpu suite)
  - 13 test cases covering all functionality

Commit:
Merge AutoWeightsLoader test finalization

Commit:
Revert "Test AutoWeightsLoader with Qwen2 model"

This reverts commit 8e1353cf5f080aba40caa32396885cc03cd57598.

Commit:
Added tests for AutoWeightLoader

adityakamat24 commented:
@yhyang201 

Hey, sorry for being a bit messy.

 I've added comprehensive unit tests for `AutoWeightsLoader`:

  - **13 test cases** covering all functionality (`test/srt/test_auto_weights_loader.py`)
  - **Integrated into CI** - Added to `per-commit-1-gpu` suite in `run_suite.py` (line 69)
  - **Fast execution** - Estimated runtime: ~5 seconds
  - **Qwen2 integration reverted** - Keeping the utility separate from example usage

  **Test coverage:**
  - Basic weight loading
  - Skip rules (prefixes, substrings)
  - Nested modules
  - Custom weight loaders
  - Ignore rules for unexpected weights
  - Edge cases (empty weights, defensive copying)
  - Rotary embedding auto-skip

  All tests pass locally. The unit tests provide sufficient CI coverage.

  Let me know if you'd like any changes or additional test cases!
